### PR TITLE
provider/aws: Add support for iam_roles to aws_rds_cluster

### LINKED ISSUE #10166: Feature Request: Specify IAM role(s) for RDS Cluster
AWS Aurora [recently](https://aws.amazon.com/about-aws/whats-new/2016/10/amazon-aurora-new-features-aws-lambda-integration-and-data-load-from-amazon-s3-to-aurora-tables/) added supports adding roles to the RDS cluster. This can be used to integrate Aurora with other AWS services, such as S3 or Lambda.

For details on this please see: http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Aurora.Authorizing.AWSServices.html

It would be nice to have the ability to add IAM roles to an an `aws_rds_cluster` in much the same way as the `aws_redshift_cluster` resource's `iam_roles` attribute.

### PR DESCRIPTION
**Pull request by stack72**
Fixes #10166

### CODE REVIEW

**Comment by willejs** [comment]
@stack72 looks like this works, any reason it cant get merged? Is it because its missing tests?

**Comment by stack72** [comment]
Hi @willejs 

I am actually working on this code *right* now to get it finished :) Stick with me for a little while - it *will* be in the next version of TF

P.

File: builtin/providers/aws/resource_aws_rds_cluster.go [hunk 1]
```diff
@@ -499,6 +511,16 @@ func resourceAwsRDSClusterRead(d *schema.ResourceData, meta interface{}) error {
 	d.Set("reader_endpoint", dbc.ReaderEndpoint)
 	d.Set("replication_source_identifier", dbc.ReplicationSourceIdentifier)
 
+	var roles []string
+	for _, r := range dbc.AssociatedRoles {
+		if *r.Status == "ACTIVE" {
+			roles = append(roles, *r.RoleArn)
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_rds_cluster.go]: The role status is one of `ACTIVE`, `PENDING`, or `INVALID`; do you not think it's worth storing the role even if it's not in the right state? I worry something will be `INVALID` and the user won't see it, and then Terraform will think it's just not there and try to reapply it.

File: builtin/providers/aws/resource_aws_rds_cluster.go [hunk 2]
```diff
@@ -579,6 +601,27 @@ func resourceAwsRDSClusterUpdate(d *schema.ResourceData, meta interface{}) error
 		}
 	}
 
+	if d.HasChange("iam_roles") {
+		oraw, nraw := d.GetChange("iam_roles")
+		o := oraw.([]interface{})
+		n := nraw.([]interface{})
+		create, remove := rdsClusterDiffIamRoles(o, n)
+
+		for _, role := range create {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_rds_cluster.go]: Are you aware of any reason where 2 roles could conflict, such that adding the new ones before removing the old ones, could cause a conflict or failure here?

File: builtin/providers/aws/resource_aws_rds_cluster.go (hunk 1 above)
Comment (stack72) [inline builtin/providers/aws/resource_aws_rds_cluster.go]: So the issue I faced here is that the UI only shows ACTIVE roles but the API returns the full history of the role associations - so i decided that we would only grab the attached ones

Maybe this isn't correct?

P.

File: builtin/providers/aws/resource_aws_rds_cluster.go (hunk 2 above)
Comment (stack72) [inline builtin/providers/aws/resource_aws_rds_cluster.go]: No, they have to be uniquely named - they cannot attach the same role twice

File: builtin/providers/aws/resource_aws_rds_cluster.go (hunk 1 above)
Comment (grubernaut) [inline builtin/providers/aws/resource_aws_rds_cluster.go]: I would have `roles` be a map or a set, so the user could query what the `STATUS` is of a role. It looks to me, that if a role is in the `INVALID` state, it's still attached to the db cluster, preventing that role from being attached again.

File: builtin/providers/aws/resource_aws_rds_cluster_test.go [hunk 3]
```diff
@@ -557,6 +557,71 @@ resource "aws_rds_cluster" "default" {
     Environment = "production"
   }
 }`, n, n, n)
+
```
Comment (mikecook) [inline builtin/providers/aws/resource_aws_rds_cluster_test.go]: missing ```}``` here

Comment (mikecook) [inline builtin/providers/aws/resource_aws_rds_cluster_test.go]: Here's the diff...

```diff --git a/builtin/providers/aws/resource_aws_rds_cluster_test.go b/builtin/providers/aws/resource_aws_rds_cluster_test.go
index 798db9c4c..d73a0697c 100644
--- a/builtin/providers/aws/resource_aws_rds_cluster_test.go
+++ b/builtin/providers/aws/resource_aws_rds_cluster_test.go
@@ -557,6 +557,7 @@ resource "aws_rds_cluster" "default" {
     Environment = "production"
   }
 }`, n, n, n)
+}
 
 func testAccAWSClusterConfig_namePrefix(n int) string {
        return fmt.Sprintf(`
```

**Comment by stack72** [comment]
Closed as migrating to new repository

**Pull request closed without merging**